frappe/modules/import_file.py

```diff
@@ -44,44 +62,105 @@ def get_file_path(module, dt, dn):
+		If the return is `none,` this file doesn't exist in the DB, so Import the file. If not, move ahead.
+	- Check if there is a hash in DB for that file. If there is, Calculate the Hash of the file to import and compare it with the one in DB if they are not equal.
+		Import the file. If Hash doesn't exist, move ahead.
+	- Check if `db_modified_timestamp` is older than the timestamp in the file; if it is, we import the file.
+
+	If timestamp comparison happens for doctypes, that means the Hash for it doesn't exist.
+	So, even if the timestamp is newer on DB (When comparing timestamps), we import the file and add the calculated Hash to the DB.
+	So in the subsequent imports, we can use hashes to compare. As a precautionary measure, the timestamp is updated to the current time as well.
+
+	Args:
+		path (str): Path to the file.
+		force (bool, optional): Load the file without checking any conditions. Defaults to False.
+		data_import (bool, optional): [description]. Defaults to False.
+		pre_process ([type], optional): Any preprocesing that may need to take place on the doc. Defaults to None.
+		ignore_version (bool, optional): ignore current version. Defaults to None.
+		reset_permissions (bool, optional): reset permissions for the file. Defaults to False.
+
+	Returns:
+		[bool]: True if import takes place. False if it wasn't imported.
+	"""
+	frappe.flags.dt = frappe.flags.dt or []
 	try:
 		docs = read_doc_from_file(path)
 	except IOError:
 		print (path + " missing")
 		return
 
+	calculated_hash = caclulate_hash(path)
+
 	if docs:
 		if not isinstance(docs, list):
 			docs = [docs]
 
 		for doc in docs:
-			if not force:
-				# check if timestamps match
-				db_modified = frappe.db.get_value(doc['doctype'], doc['name'], 'modified')
-				if db_modified and doc.get('modified')==get_datetime_str(db_modified):
-					return False
 
-			original_modified = doc.get("modified")
+			# modified timestamp in db, none if doctype's first import
+			db_modified_timestamp = frappe.db.get_value(doc["doctype"], doc["name"], "modified")
+			is_db_timestamp_latest = db_modified_timestamp and doc.get("modified") <= get_datetime_str(db_modified_timestamp)
+
+			if not force or db_modified_timestamp:
+				try:
+					stored_hash = frappe.db.get_value(doc["doctype"], doc["name"], "migration_hash")
+				except Exception:
+					frappe.flags.dt += [doc["doctype"]]
+					stored_hash = None
+
+				# if hash exists and is equal no need to update
+				if stored_hash and stored_hash == calculated_hash:
+					return False
 
-			frappe.flags.in_import = True
-			import_doc(doc, force=force, data_import=data_import, pre_process=pre_process,
-				ignore_version=ignore_version, reset_permissions=reset_permissions)
-			frappe.flags.in_import = False
+				# if hash doesn't exist, check if db timestamp is same as json timestamp, add hash if from doctype
+				if is_db_timestamp_latest and doc["doctype"] != "DocType":
+					return False
```

Comment:
You `return` if _a single doc_ doesn't match your condition, however there are many docs in this loop, that won't get considered at afterwards. `continue` would be appropriate here.